<a href="https://colab.research.google.com/github/nephelim74/TransferLearning/blob/main/%D0%94%D0%BE%D0%BC%D0%B0%D1%88%D0%BD%D0%B5%D0%B5_%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B53.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнее задание

Реализуйте классификацию эмоциональной окрашенности текстов при помощи объекта pipeline.

1. Найдите тип задач Sentiment Analysis на huggingface.co
2. Найдите модель для русского языка (примеры: rubert-tiny2..., rubert-base...)4

import transformers

classifier = pipeline( ... )

classifier( ["Отличный товар",
            "Мне не понравилось",
            "Прекрасно выглядит, ничего не меняйте!",
            "Ничего не понял",
            "У меня не работала розетка, в остальном все отлично"] )



In [1]:
# Для удобства загрузим модели на локальный диск:
from transformers import AutoModelForSequenceClassification, AutoTokenizer
# модель: rubert-tiny2-russian-sentiment
model_name = "seara/rubert-tiny2-russian-sentiment"
# Загружаем модель и токенизатор
model = AutoModelForSequenceClassification.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name)
# Сохраняем модель и токенизатор локально
model.save_pretrained("./rubert-tiny2-sentiment")
tokenizer.save_pretrained("./rubert-tiny2-sentiment")
# модель: rubert-base-cased-russian-sentiment
model_name = "seara/rubert-base-cased-russian-sentiment"
# Загружаем модель и токенизатор
model = AutoModelForSequenceClassification.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name)
# Сохраняем модель и токенизатор локально
model.save_pretrained("./rubert-base-cased-russian-sentiment")
tokenizer.save_pretrained("./rubert-base-cased-russian-sentiment")


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/911 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/117M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/368 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/1.08M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.41M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.08k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/711M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/395 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/1.65M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/3.57M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

('./rubert-base-cased-russian-sentiment/tokenizer_config.json',
 './rubert-base-cased-russian-sentiment/special_tokens_map.json',
 './rubert-base-cased-russian-sentiment/vocab.txt',
 './rubert-base-cased-russian-sentiment/added_tokens.json',
 './rubert-base-cased-russian-sentiment/tokenizer.json')

In [3]:
from transformers import pipeline

In [9]:
# Загружаем модель и токенизатор из локальной папки, для удобства делаем это через функцию
def f_model(token_name, path):
  sentiment_pipeline = pipeline(token_name, model=path)
  return sentiment_pipeline


In [10]:
## Поэкспериментируем с моделью rubert-tiny2-russian-sentiment
classifier = f_model("sentiment-analysis", "./rubert-tiny2-sentiment")

In [11]:
classifier( ["Отличный товар",
            "Мне не понравилось",
            "Прекрасно выглядит, ничего не меняйте!",
            "Ничего не понял",
            "У меня не работала розетка, в остальном все отлично",] )

[{'label': 'positive', 'score': 0.9941100478172302},
 {'label': 'neutral', 'score': 0.5911950469017029},
 {'label': 'positive', 'score': 0.9389205574989319},
 {'label': 'negative', 'score': 0.524005651473999},
 {'label': 'positive', 'score': 0.6734204888343811}]

In [12]:
## Поэкспериментируем с моделью rubert-tiny2-russian-sentiment
classifier = f_model("sentiment-analysis", "./rubert-base-cased-russian-sentiment")

In [13]:
classifier( ["Отличный товар",
            "Мне не понравилось",
            "Прекрасно выглядит, ничего не меняйте!",
            "Ничего не понял",
            "У меня не работала розетка, в остальном все отлично"] )

[{'label': 'positive', 'score': 0.9960761666297913},
 {'label': 'neutral', 'score': 0.701276421546936},
 {'label': 'positive', 'score': 0.9847234487533569},
 {'label': 'neutral', 'score': 0.6664650440216064},
 {'label': 'positive', 'score': 0.939444363117218}]

### Как видим мождели предобучены несколько по разному. Если в rubert-tiny2-sentiment на фразу "Ничего не понял", реакция 'negative'с вероятностью 0.524005651473999, то в rubert-base-cased-russian-sentiment это neutral с вероятностью 0.6664650440216064. Если все же здесь речь идет об отзывах о товарах, то наверное обе модели потребуется доучить. В текущем контексте фразу "Ничего не понял" все же следует рассматривать как нейтральную. То же самое касаемо фразы "У меня не работала розетка, в остальном все отлично". Все же доля негатитва, из-за того, что не работала розетка привнесено. Но и та и другая модель определили эту фразу как positive. Толко  первая с вероятностью 0.6734204888343811, а другая 0.939444363117218. Все же у первой доля сомнения больше. Продолжим экспериментировать. По идее фраза "У меня все не работает" должна интерпретироваться как negative. Посмотрим, как справяться с этой фразом модели.

In [19]:
classifier = f_model("sentiment-analysis", "./rubert-tiny2-sentiment")
classifier( ["У меня все не работает"])

[{'label': 'negative', 'score': 0.5317866802215576}]

In [20]:
classifier = f_model("sentiment-analysis", "./rubert-base-cased-russian-sentiment")
classifier( ["У меня все не работает"])

[{'label': 'neutral', 'score': 0.4993921220302582}]

# Как видим с небольшой долей вероятности с этим справилась модель rubert-tiny2-sentiment. Из этого следует, что данная модель больше подходит для дальнейшего использования для "натаскивания их" для оценок торговых площадок